### PR TITLE
[Dy2Stat] Add cache for Executor and Context in run_program_op

### PR DESCRIPTION
**Pull request by Aurelius84**
### PR types

Performance optimization
### PR changes

OPs
### Describe

| 模型                | 动态图    | to_static | to_static (cache)   |
| ------------------- | --------- | --------- | ------------------- |
| ResNet (`ms`)       | **60.3**  | 91.1      | 62.4 (`+45%`)       |
| SeResNet (`step/s`) | **3.618** | 2.925     | 3.40 (`+16%`)       |
| Ptb LM (`step/s`)   | 180       | 82        | **222** (`+170%` )  |
| Seq2Seq (`ms`)      | **41.7**  | 552       | 540 (`+2.2%`)       |
| BMN (`ms`)          | **28.2**  | 41.3      | 33.1 (`+24.1%`)     |
| Cycle GAN (`ms`)    | 90        | 135.4     | **55 (`+146%`)**    |
| BERT (`step/s`)     | 10.7      | 12.05     | **13.12 (`+8.8%`)** |
| MobileNet V1 (`ms`) | **28.7**  | 45.6      | 31.9 (`+42.9%`)     |
| MobileNet V2 (`ms`) | 58.6      | 74.0      | **50.2 (`+47.4%`)** |

### CODE REVIEW

**Review by liym27 — DISMISSED** [review]
LGTM

File: paddle/fluid/operators/run_program_op.h [hunk 1]
```diff
@@ -196,10 +202,45 @@ static void AppendSafeEagerDeletionSkipVars(
...
+  }
+
+  return cached_exe_info->at(program);
+}
+
 }  // namespace details
 
 template <typename DeviceContext, typename T>
 class RunProgramOpKernel : public framework::OpKernel<T> {
+ private:
```
Comment (chenwhql) [inline paddle/fluid/operators/run_program_op.h]: by google cpp style, put `private` after `public`
image

File: paddle/fluid/operators/run_program_op.h [hunk 2]
```diff
@@ -196,10 +202,45 @@ static void AppendSafeEagerDeletionSkipVars(
...
+  static std::unordered_map<framework::ProgramDesc *, ExecutorInfo>
```
Comment (chenwhql) [inline paddle/fluid/operators/run_program_op.h]: `static` may introduce problems,maybe singleton is better
image

File: paddle/fluid/operators/run_program_op.h [hunk 3]
```diff
@@ -278,8 +313,16 @@ class RunProgramOpKernel : public framework::OpKernel<T> {
   }
 };
 
+template <typename DeviceContext, typename T>
+std::unordered_map<framework::ProgramDesc *, ExecutorInfo>
+    RunProgramOpKernel<DeviceContext, T>::cached_exe_info_;
+
 template <typename DeviceContext, typename T>
 class RunProgramGradOpKernel : public framework::OpKernel<T> {
+ private:
```
Comment (chenwhql) [inline paddle/fluid/operators/run_program_op.h]: same above

File: paddle/fluid/operators/run_program_op.h [hunk 4]
```diff
@@ -44,6 +47,9 @@ using BlockDesc = framework::BlockDesc;
 using Variable = framework::Variable;
 using LoDTensor = framework::LoDTensor;
 using SelectedRows = framework::SelectedRows;
+using ExecutorInfo =
+    std::tuple<std::shared_ptr<framework::Executor>,
```
Comment (chenwhql) [inline paddle/fluid/operators/run_program_op.h]: maybe we only need one Executor? Executor is stateless and should be able to execute multiple ExecutorPrepareContext

Comment (Aurelius84) [reply paddle/fluid/operators/run_program_op.h]: Thanks, done

File: paddle/fluid/operators/run_program_op.h (hunk 2 above)
Comment (Aurelius84) [reply paddle/fluid/operators/run_program_op.h]: Thanks advice~~, modified into singleton

File: paddle/fluid/operators/run_program_op.h (hunk 1 above)
Comment (Aurelius84) [reply paddle/fluid/operators/run_program_op.h]: Thanks, fixed.

**Review by chenwhql — APPROVED** [review]
LGTM

**Review by liym27 — APPROVED** [review]

**Review by zhhsplendid — APPROVED** [review]
LGTM

**Pull request merged**